In [8]:
from __future__ import print_function
import numpy as np
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers.core import Dense, Dropout, Activation
from keras.optimizers import SGD
from keras.utils import np_utils
np.random.seed(1671) # for reproducibility
# network and training
NB_EPOCH = 250
BATCH_SIZE = 350
VERBOSE = 1
NB_CLASSES = 10 # number of outputs = number of digits
OPTIMIZER = SGD() # optimizer, explained later in this chapter
N_HIDDEN = 2
VALIDATION_SPLIT=0.2 # how much TRAIN is reserved for VALIDATION
DROPOUT = 0.3
# data: shuffled and split between train and test sets
(X_train, y_train), (X_test, y_test) = mnist.load_data()
#X_train is 60000 rows of 28x28 values --> reshaped in 60000 x 784
RESHAPED = 784
#
X_train = X_train.reshape(60000, RESHAPED)
X_test = X_test.reshape(10000, RESHAPED)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
# normalize
X_train /= 255
X_test /= 255
# convert class vectors to binary class matrices
Y_train = np_utils.to_categorical(y_train, NB_CLASSES)
Y_test = np_utils.to_categorical(y_test, NB_CLASSES)
# M_HIDDEN hidden layers 10 outputs
model = Sequential()
model.add(Dense(N_HIDDEN, input_shape=(RESHAPED,)))
model.add(Activation('relu'))
model.add(Dropout(DROPOUT))
model.add(Dense(N_HIDDEN))
model.add(Activation('relu'))
model.add(Dropout(DROPOUT))
model.add(Dense(NB_CLASSES))
model.add(Activation('softmax'))
model.summary()
model.compile(loss='categorical_crossentropy',
optimizer=OPTIMIZER,
metrics=['accuracy'])
history = model.fit(X_train, Y_train,
batch_size=BATCH_SIZE, epochs=NB_EPOCH,
verbose=VERBOSE, validation_split=VALIDATION_SPLIT)
score = model.evaluate(X_test, Y_test, verbose=VERBOSE)
print("Test score:", score[0])
print('Test accuracy:', score[1])

Model: "sequential_6"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_18 (Dense)            (None, 2)                 1570      
                                                                 
 activation_18 (Activation)  (None, 2)                 0         
                                                                 
 dropout_12 (Dropout)        (None, 2)                 0         
                                                                 
 dense_19 (Dense)            (None, 2)                 6         
                                                                 
 activation_19 (Activation)  (None, 2)                 0         
                                                                 
 dropout_13 (Dropout)        (None, 2)                 0         
                                                                 
 dense_20 (Dense)            (None, 10)               

Epoch 48/250
138/138 [==============================] - 0s 1ms/step - loss: 2.1307 - accuracy: 0.1734 - val_loss: 1.9714 - val_accuracy: 0.2705
Epoch 49/250
138/138 [==============================] - 0s 1ms/step - loss: 2.1324 - accuracy: 0.1720 - val_loss: 1.9734 - val_accuracy: 0.2724
Epoch 50/250
138/138 [==============================] - 0s 1ms/step - loss: 2.1321 - accuracy: 0.1717 - val_loss: 1.9660 - val_accuracy: 0.2721
Epoch 51/250
138/138 [==============================] - 0s 1ms/step - loss: 2.1326 - accuracy: 0.1720 - val_loss: 1.9679 - val_accuracy: 0.2751
Epoch 52/250
138/138 [==============================] - 0s 1ms/step - loss: 2.1313 - accuracy: 0.1719 - val_loss: 1.9603 - val_accuracy: 0.2746
Epoch 53/250
138/138 [==============================] - 0s 1ms/step - loss: 2.1297 - accuracy: 0.1731 - val_loss: 1.9615 - val_accuracy: 0.2777
Epoch 54/250
138/138 [==============================] - 0s 1ms/step - loss: 2.1238 - accuracy: 0.1739 - val_loss: 1.9567 - val_accuracy:

Epoch 105/250
138/138 [==============================] - 0s 1ms/step - loss: 2.1233 - accuracy: 0.1738 - val_loss: 1.9372 - val_accuracy: 0.2851
Epoch 106/250
138/138 [==============================] - 0s 1ms/step - loss: 2.1203 - accuracy: 0.1755 - val_loss: 1.9272 - val_accuracy: 0.2845
Epoch 107/250
138/138 [==============================] - 0s 1ms/step - loss: 2.1171 - accuracy: 0.1759 - val_loss: 1.9306 - val_accuracy: 0.2834
Epoch 108/250
138/138 [==============================] - 0s 1ms/step - loss: 2.1198 - accuracy: 0.1755 - val_loss: 1.9306 - val_accuracy: 0.2842
Epoch 109/250
138/138 [==============================] - 0s 1ms/step - loss: 2.1222 - accuracy: 0.1744 - val_loss: 1.9326 - val_accuracy: 0.2882
Epoch 110/250
138/138 [==============================] - 0s 1ms/step - loss: 2.1194 - accuracy: 0.1750 - val_loss: 1.9334 - val_accuracy: 0.2845
Epoch 111/250
138/138 [==============================] - 0s 1ms/step - loss: 2.1183 - accuracy: 0.1759 - val_loss: 1.9348 - val_ac

138/138 [==============================] - 0s 1ms/step - loss: 2.1115 - accuracy: 0.1769 - val_loss: 1.9278 - val_accuracy: 0.2890
Epoch 218/250
138/138 [==============================] - 0s 1ms/step - loss: 2.1155 - accuracy: 0.1766 - val_loss: 1.9263 - val_accuracy: 0.2892
Epoch 219/250
138/138 [==============================] - 0s 1ms/step - loss: 2.1128 - accuracy: 0.1765 - val_loss: 1.9231 - val_accuracy: 0.2866
Epoch 220/250
138/138 [==============================] - 0s 1ms/step - loss: 2.1134 - accuracy: 0.1767 - val_loss: 1.9233 - val_accuracy: 0.2846
Epoch 221/250
138/138 [==============================] - 0s 1ms/step - loss: 2.1184 - accuracy: 0.1752 - val_loss: 1.9258 - val_accuracy: 0.2856
Epoch 222/250
138/138 [==============================] - 0s 1ms/step - loss: 2.1121 - accuracy: 0.1772 - val_loss: 1.9173 - val_accuracy: 0.2878
Epoch 223/250
138/138 [==============================] - 0s 1ms/step - loss: 2.1159 - accuracy: 0.1759 - val_loss: 1.9290 - val_accuracy: 0.2842

In [7]:
from __future__ import print_function
import numpy as np
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers.core import Dense, Dropout, Activation
from keras.optimizers import SGD
from keras.utils import np_utils
np.random.seed(1671) # for reproducibility
# network and training
NB_EPOCH = 250
BATCH_SIZE = 350
VERBOSE = 1
NB_CLASSES = 10 # number of outputs = number of digits
OPTIMIZER = SGD() # optimizer, explained later in this chapter
N_HIDDEN = 128
VALIDATION_SPLIT=0.2 # how much TRAIN is reserved for VALIDATION
DROPOUT = 0.3
# data: shuffled and split between train and test sets
(X_train, y_train), (X_test, y_test) = mnist.load_data()
#X_train is 60000 rows of 28x28 values --> reshaped in 60000 x 784
RESHAPED = 784
#
X_train = X_train.reshape(60000, RESHAPED)
X_test = X_test.reshape(10000, RESHAPED)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
# normalize
X_train /= 255
X_test /= 255
# convert class vectors to binary class matrices
Y_train = np_utils.to_categorical(y_train, NB_CLASSES)
Y_test = np_utils.to_categorical(y_test, NB_CLASSES)
# M_HIDDEN hidden layers 10 outputs
model = Sequential()
model.add(Dense(N_HIDDEN, input_shape=(RESHAPED,)))
model.add(Activation('relu'))
model.add(Dropout(DROPOUT))
model.add(Dense(N_HIDDEN))
model.add(Activation('relu'))
model.add(Dropout(DROPOUT))
model.add(Dense(NB_CLASSES))
model.add(Activation('softmax'))
model.summary()
model.compile(loss='categorical_crossentropy',
optimizer=OPTIMIZER,
metrics=['accuracy'])
history = model.fit(X_train, Y_train,
batch_size=BATCH_SIZE, epochs=NB_EPOCH,
verbose=VERBOSE, validation_split=VALIDATION_SPLIT)
score = model.evaluate(X_test, Y_test, verbose=VERBOSE)
print("Test score:", score[0])
print('Test accuracy:', score[1])

Model: "sequential_5"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_15 (Dense)            (None, 128)               100480    
                                                                 
 activation_15 (Activation)  (None, 128)               0         
                                                                 
 dropout_10 (Dropout)        (None, 128)               0         
                                                                 
 dense_16 (Dense)            (None, 128)               16512     
                                                                 
 activation_16 (Activation)  (None, 128)               0         
                                                                 
 dropout_11 (Dropout)        (None, 128)               0         
                                                                 
 dense_17 (Dense)            (None, 10)               

Epoch 48/250
138/138 [==============================] - 0s 3ms/step - loss: 0.3047 - accuracy: 0.9108 - val_loss: 0.2036 - val_accuracy: 0.9407
Epoch 49/250
138/138 [==============================] - 0s 3ms/step - loss: 0.2963 - accuracy: 0.9137 - val_loss: 0.2017 - val_accuracy: 0.9413
Epoch 50/250
138/138 [==============================] - 0s 3ms/step - loss: 0.2949 - accuracy: 0.9137 - val_loss: 0.1998 - val_accuracy: 0.9421
Epoch 51/250
138/138 [==============================] - 0s 3ms/step - loss: 0.2972 - accuracy: 0.9133 - val_loss: 0.1990 - val_accuracy: 0.9416
Epoch 52/250
138/138 [==============================] - 0s 3ms/step - loss: 0.2933 - accuracy: 0.9133 - val_loss: 0.1967 - val_accuracy: 0.9435
Epoch 53/250
138/138 [==============================] - 0s 3ms/step - loss: 0.2893 - accuracy: 0.9155 - val_loss: 0.1946 - val_accuracy: 0.9438
Epoch 54/250
138/138 [==============================] - 0s 3ms/step - loss: 0.2840 - accuracy: 0.9168 - val_loss: 0.1934 - val_accuracy:

Epoch 105/250
138/138 [==============================] - 0s 3ms/step - loss: 0.2065 - accuracy: 0.9389 - val_loss: 0.1417 - val_accuracy: 0.9587
Epoch 106/250
138/138 [==============================] - 0s 3ms/step - loss: 0.2073 - accuracy: 0.9387 - val_loss: 0.1414 - val_accuracy: 0.9591
Epoch 107/250
138/138 [==============================] - 0s 3ms/step - loss: 0.2066 - accuracy: 0.9398 - val_loss: 0.1410 - val_accuracy: 0.9594
Epoch 108/250
138/138 [==============================] - 0s 3ms/step - loss: 0.2042 - accuracy: 0.9406 - val_loss: 0.1400 - val_accuracy: 0.9591
Epoch 109/250
138/138 [==============================] - 0s 3ms/step - loss: 0.2026 - accuracy: 0.9407 - val_loss: 0.1397 - val_accuracy: 0.9592
Epoch 110/250
138/138 [==============================] - 0s 3ms/step - loss: 0.2016 - accuracy: 0.9406 - val_loss: 0.1389 - val_accuracy: 0.9598
Epoch 111/250
138/138 [==============================] - 0s 3ms/step - loss: 0.2011 - accuracy: 0.9410 - val_loss: 0.1389 - val_ac

138/138 [==============================] - 0s 3ms/step - loss: 0.1364 - accuracy: 0.9600 - val_loss: 0.1030 - val_accuracy: 0.9697
Epoch 218/250
138/138 [==============================] - 0s 3ms/step - loss: 0.1356 - accuracy: 0.9598 - val_loss: 0.1029 - val_accuracy: 0.9700
Epoch 219/250
138/138 [==============================] - 0s 3ms/step - loss: 0.1346 - accuracy: 0.9592 - val_loss: 0.1025 - val_accuracy: 0.9699
Epoch 220/250
138/138 [==============================] - 0s 3ms/step - loss: 0.1347 - accuracy: 0.9606 - val_loss: 0.1022 - val_accuracy: 0.9701
Epoch 221/250
138/138 [==============================] - 0s 3ms/step - loss: 0.1336 - accuracy: 0.9603 - val_loss: 0.1024 - val_accuracy: 0.9699
Epoch 222/250
138/138 [==============================] - 0s 3ms/step - loss: 0.1340 - accuracy: 0.9598 - val_loss: 0.1019 - val_accuracy: 0.9702
Epoch 223/250
138/138 [==============================] - 0s 3ms/step - loss: 0.1377 - accuracy: 0.9585 - val_loss: 0.1015 - val_accuracy: 0.9703

In [6]:
from __future__ import print_function
import numpy as np
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers.core import Dense, Dropout, Activation
from keras.optimizers import SGD
from keras.utils import np_utils
np.random.seed(1671) # for reproducibility
# network and training
NB_EPOCH = 250
BATCH_SIZE = 350
VERBOSE = 1
NB_CLASSES = 10 # number of outputs = number of digits
OPTIMIZER = SGD() # optimizer, explained later in this chapter
N_HIDDEN = 256
VALIDATION_SPLIT=0.01 # how much TRAIN is reserved for VALIDATION
DROPOUT = 0.3
# data: shuffled and split between train and test sets
(X_train, y_train), (X_test, y_test) = mnist.load_data()
#X_train is 60000 rows of 28x28 values --> reshaped in 60000 x 784
RESHAPED = 784
#
X_train = X_train.reshape(60000, RESHAPED)
X_test = X_test.reshape(10000, RESHAPED)
X_train = X_train.astype('float32')
X_test = X_test.astype('float32')
# normalize
X_train /= 255
X_test /= 255
# convert class vectors to binary class matrices
Y_train = np_utils.to_categorical(y_train, NB_CLASSES)
Y_test = np_utils.to_categorical(y_test, NB_CLASSES)
# M_HIDDEN hidden layers 10 outputs
model = Sequential()
model.add(Dense(N_HIDDEN, input_shape=(RESHAPED,)))
model.add(Activation('relu'))
model.add(Dropout(DROPOUT))
model.add(Dense(N_HIDDEN))
model.add(Activation('relu'))
model.add(Dropout(DROPOUT))
model.add(Dense(NB_CLASSES))
model.add(Activation('softmax'))
model.summary()
model.compile(loss='categorical_crossentropy',
optimizer=OPTIMIZER,
metrics=['accuracy'])
history = model.fit(X_train, Y_train,
batch_size=BATCH_SIZE, epochs=NB_EPOCH,
verbose=VERBOSE, validation_split=VALIDATION_SPLIT)
score = model.evaluate(X_test, Y_test, verbose=VERBOSE)
print("Test score:", score[0])
print('Test accuracy:', score[1])

Model: "sequential_4"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_12 (Dense)            (None, 256)               200960    
                                                                 
 activation_12 (Activation)  (None, 256)               0         
                                                                 
 dropout_8 (Dropout)         (None, 256)               0         
                                                                 
 dense_13 (Dense)            (None, 256)               65792     
                                                                 
 activation_13 (Activation)  (None, 256)               0         
                                                                 
 dropout_9 (Dropout)         (None, 256)               0         
                                                                 
 dense_14 (Dense)            (None, 10)               

Epoch 48/250
170/170 [==============================] - 1s 5ms/step - loss: 0.2297 - accuracy: 0.9331 - val_loss: 0.1632 - val_accuracy: 0.9650
Epoch 49/250
170/170 [==============================] - 1s 5ms/step - loss: 0.2261 - accuracy: 0.9331 - val_loss: 0.1626 - val_accuracy: 0.9650
Epoch 50/250
170/170 [==============================] - 1s 5ms/step - loss: 0.2260 - accuracy: 0.9344 - val_loss: 0.1611 - val_accuracy: 0.9650
Epoch 51/250
170/170 [==============================] - 1s 4ms/step - loss: 0.2214 - accuracy: 0.9354 - val_loss: 0.1611 - val_accuracy: 0.9650
Epoch 52/250
170/170 [==============================] - 1s 4ms/step - loss: 0.2210 - accuracy: 0.9347 - val_loss: 0.1596 - val_accuracy: 0.9650
Epoch 53/250
170/170 [==============================] - 1s 4ms/step - loss: 0.2171 - accuracy: 0.9359 - val_loss: 0.1595 - val_accuracy: 0.9650
Epoch 54/250
170/170 [==============================] - 1s 5ms/step - loss: 0.2170 - accuracy: 0.9360 - val_loss: 0.1589 - val_accuracy:

Epoch 105/250
170/170 [==============================] - 1s 4ms/step - loss: 0.1478 - accuracy: 0.9563 - val_loss: 0.1404 - val_accuracy: 0.9733
Epoch 106/250
170/170 [==============================] - 1s 4ms/step - loss: 0.1443 - accuracy: 0.9572 - val_loss: 0.1409 - val_accuracy: 0.9733
Epoch 107/250
170/170 [==============================] - 1s 4ms/step - loss: 0.1457 - accuracy: 0.9562 - val_loss: 0.1407 - val_accuracy: 0.9733
Epoch 108/250
170/170 [==============================] - 1s 4ms/step - loss: 0.1442 - accuracy: 0.9570 - val_loss: 0.1397 - val_accuracy: 0.9733
Epoch 109/250
170/170 [==============================] - 1s 4ms/step - loss: 0.1402 - accuracy: 0.9584 - val_loss: 0.1406 - val_accuracy: 0.9733
Epoch 110/250
170/170 [==============================] - 1s 4ms/step - loss: 0.1403 - accuracy: 0.9581 - val_loss: 0.1401 - val_accuracy: 0.9733
Epoch 111/250
170/170 [==============================] - 1s 4ms/step - loss: 0.1398 - accuracy: 0.9582 - val_loss: 0.1399 - val_ac

170/170 [==============================] - 1s 4ms/step - loss: 0.0902 - accuracy: 0.9730 - val_loss: 0.1359 - val_accuracy: 0.9783
Epoch 218/250
170/170 [==============================] - 1s 4ms/step - loss: 0.0897 - accuracy: 0.9726 - val_loss: 0.1356 - val_accuracy: 0.9783
Epoch 219/250
170/170 [==============================] - 1s 4ms/step - loss: 0.0900 - accuracy: 0.9734 - val_loss: 0.1355 - val_accuracy: 0.9767
Epoch 220/250
170/170 [==============================] - 1s 4ms/step - loss: 0.0876 - accuracy: 0.9735 - val_loss: 0.1354 - val_accuracy: 0.9767
Epoch 221/250
170/170 [==============================] - 1s 4ms/step - loss: 0.0876 - accuracy: 0.9732 - val_loss: 0.1358 - val_accuracy: 0.9783
Epoch 222/250
170/170 [==============================] - 1s 4ms/step - loss: 0.0887 - accuracy: 0.9733 - val_loss: 0.1351 - val_accuracy: 0.9767
Epoch 223/250
170/170 [==============================] - 1s 4ms/step - loss: 0.0889 - accuracy: 0.9728 - val_loss: 0.1356 - val_accuracy: 0.9767

To establish a accuracy rate for training I used the test data sets with two hidden layers comfirm that the code was working as intented. Than I changed the number of hidden partitions in order to set different peramiters with in the code allowing me to see what parameters would overal chance. There parameters got significantly larger with the density almost trippling in size. There was also a change to the validation split from .2 to .01 that will overall affect the accuracy of the tests. These changes has increased the peramiters by 44% making the overall perameters larger. The second test had a accuracy score of 97% and test score of .074 while first test score of .094 with 96% accuracy. With the changes to the parameters we can see a large differance between the two test with the over all score values and how this made changes to the accuracy of the test.